src/optimizer_interface.jl

```diff
@@ -573,6 +573,65 @@ function dual_status(model::GenericModel; result::Int = 1)
     return MOI.get(model, MOI.DualStatus(result))::MOI.ResultStatusCode
 end
 
+"""
+    has_optimal_solution(
+        model::GenericModel;
+        dual::Bool = false,
+        allow_local::Bool = true,
+        allow_almost::Bool = false,
+        result::Int = 1,
+    )
+
+Return `true` if the model has an optimal primal solution associated with result
+index `result`.
+
+If `dual`, additionally check that an optimal dual solution is available.
+
+If `allow_local`, the [`termination_status`](@ref) may be [`OPTIMAL`](@ref) (the
+solver found a global optimum) or [`LOCALLY_SOLVED`](@ref) (the solver found a
+local optimum, which may also be the global optimum, but the solver could not
+prove so).
+
+If `allow_local = false`, then this function returns `true` only if the
+[`termination_status`](@ref) is [`OPTIMAL`](@ref).
+
+If `allow_almost = true`, then the [`termination_status`](@ref) may additionally
+be [`ALMOST_OPTIMAL`](@ref) or [`ALMOST_LOCALLY_SOLVED`](@ref) (if `allow_local`),
+and the [`primal_status`](@ref) and [`dual_status`](@ref) may additionally be
+[`NEARLY_FEASIBLE_POINT`](@ref).
+
+If this function returns `false`, use [`termination_status`](@ref),
+[`result_count`](@ref), [`primal_status`](@ref) and [`dual_status`](@ref) to
+understand what solutions are available (if any).
+"""
+function has_optimal_solution(
+    model::GenericModel;
+    dual::Bool = false,
+    allow_local::Bool = true,
```

Comment:
It's quite counter-intuitive that non-optimal solution are considered optimal by default. I would use `allow_local = false` by default.

Reply:
Hard disagree. This is an entry point function for users who don't need to know the details. They'll complain if they need to write `has_optimal_soution(model; allow_local = true)` when solving an LP with Ipopt but not with HiGHS.

"optimal" is widely used for both global and local minima. I don't really want to get into the nitty gritty. (If we did, we should probably change `allow_local` to be `allow_stationary_point`...)

Reply:
@mlubin your thoughts would be useful

Reply:
I don't think it's a technical detail. From the user that does not know anything about how the problem is solved, he will rightfully think that `has_optimal_solution` means the solution is optimal. If you solve an LP with Ipopt, Maybe Ipopt could detect the problem is convex and return `MOI.OPTIMAL` then.

Reply:
Ipopt cannot detect convexity. 

I don't think people care about the difference between global and local optima. Either problem is convex and we could a global, or the problem is non-convex, and they probably want only a local. If the problem is non-convex and they want the global, they're doing something special and can opt-out.

Even then, `AmplNLWriter` returns `LOCALLY_SOLVED` even if you use a global solver like Couenne (because we can't know what the executable at `"/path/to/random/ampl.exe"` is).

I don't want to have to explain to new users why they can use `has_optimal_solution(model)` for some solvers and `has_optimal_solution(model; allow_local = true)` for others.

Reply:
I think it's confusing to break with MOI and start saying that `LOCALLY_SOLVED` is "optimal". Can we use a different word like "solved" instead of "optimal"? `is_solved`?

Reply:
Some alternatives:

 * `has_solution(model; dual = true)`
 
 * `is_solved(model)`
 * `was_solved(model)`
 * `solved_successfully(model)`
 * `solved_successfully(model; require_global = true)` <-- an new option

 * `is_optimize_success(model)`
 * `optimized_successfully(model)`
 * `optimize_succeeded(model)`

Reply:
I think we thought carefully about MOI statuses.
So the new function should take longer to bikeshed, otherwise we might commit to something that will prove to be bad later.

I vote for fixing the entire manual to use termination_status and has_primal all around. So that people can start doing it right.

I would leave the new function for a larger and careful discussion.

Reply:
Still `termination_success(model)` might be more compatible with MOI.

Reply:
> I vote for fixing the entire manual to use termination_status and has_primal all around. So that people can start doing it right.

I did that in the first commit to this PR. But it didn't feel right. Even I wouldn't want to write that in my code.

I really want a single "all is good" call, and if not, then I can dig into the details.

I'd be okay with `is_termination_success(model)`

Reply:
`solved_successfully(model)` seems consistent with `LOCALLY_SOLVED`. It means that infeasible or unbounded is considered as unsuccessful but that's ok

Reply:
`normal_termination`?
The issue with "success" is that, from the solver's perspective, detecting infeasibility or unboundedness is a success because it did exactly what you asked it to do.
What about `ALMOST_OPTIMAL` and `ALMOST_LOCALLY_SOLVED`?

Reply:
There's an `allow_almost::Bool = false` option.

What about `has_converged_solution`, `has_solved_solution`.

Or something like
```
has_feasible_solution(
    model;
    primal::Bool = true,
    dual::Bool = false,
    allow_limit_termination::Bool = false,
    allow_local_optima::Bool = true,
    allow_almost::Bool = false,
    result::Int = 1,
)
```

Reply:
The solution could be feasible even if the status is NUMERICAL_ERROR. `has_converged_solution` is nice even though yoi could still converge to a certificate. `solved_solution` is twice `solve`. What about `is_feasibly_solved` ?

Reply:
Because it basically means that it's solved (consistent with `LOCALLY_SOLVED`) and it's feasible (consistent with the primal status). Having a function name combining these two words would make sure this is not misinterpreted

Reply:
`is_solved_and_feasible`?